cmd/intake-receiver/main.go

```diff
@@ -0,0 +1,331 @@
+		<-ctx.Done()
+		log.Println("closing http server...")
+
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		if err := srv.Shutdown(shutdownCtx); err != nil {
+			log.Println(err)
+		}
+	}()
+
+	log.Println("http server listening for requests on", srv.Addr)
+	if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+		log.Println(err)
+	}
+}
+
+type requestHandler struct {
+	agentFileMap *fileMap
+	bufPool      sync.Pool
+	bytesBufPool sync.Pool
+	basePath     string
+	rootResponse string
+}
+
+func (h requestHandler) rootHandler() http.Handler {
+	return logHandler(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/" {
+			rw.WriteHeader(200)
+			rw.Write([]byte(h.rootResponse))
+			return
+		}
+		http.Error(
+			rw,
+			fmt.Sprintf(" %s is not implemented", r.URL.Path),
+			http.StatusNotImplemented,
+		)
+	}))
+}
+
+func (h requestHandler) eventHandler() http.Handler {
+	return logHandler(
+		http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			code, err := h.handleRequest(r)
+			if err != nil {
+				log.Println("failed handling request", code, err.Error())
+				http.Error(rw, err.Error(), code)
+				return
+			}
+			rw.WriteHeader(code)
+		}),
+	)
+}
+
+func (h requestHandler) handleRequest(r *http.Request) (int, error) {
+	buf := h.bufPool.Get().(*bytes.Buffer)
+	defer func() {
+		buf.Reset()
+		h.bufPool.Put(buf)
+	}()
+
+	var err error
+	body := r.Body
+	encoding := r.Header.Get("Content-Encoding")
+	switch encoding {
+	case "deflate":
+		body, err = zlib.NewReader(r.Body)
+	case "gzip":
+		body, err = gzip.NewReader(r.Body)
+	}
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf(
+			"unable to create compressed reader for %s: %v", encoding, err,
+		)
+	}
+
+	var meta metadata
+	if err := h.processBatch(body, buf, &meta); err != nil {
+		return http.StatusBadRequest, fmt.Errorf("invalid request: %v", err)
+	}
+	if meta.IsEmpty() {
+		return http.StatusBadRequest, errors.New("agent not found in metadata")
+	}
+
+	fileName := filepath.Join(h.basePath, agentFileName(meta))
+	f, err := h.agentFileMap.Get(fileName)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf(
+			"couldn't retrieve storage file: %s", fileName,
+		)
+	}
+	h.agentFileMap.Set(fileName, f)
+
+	if _, err := io.Copy(f, buf); err != nil {
+		return http.StatusInternalServerError, fmt.Errorf(
+			"failed writing to file: %s: %v", fileName, err,
+		)
+	}
+	return http.StatusAccepted, nil
+}
+
+func (h requestHandler) processBatch(body io.ReadCloser, buf io.Writer, meta *metadata) error {
+	byteBuf := h.bytesBufPool.Get().([]byte)
+	defer func() {
+		byteBuf = byteBuf[:0]
+		h.bufPool.Put(byteBuf)
+	}()
+	defer body.Close()
+	scanner := bufio.NewScanner(body)
+	scanner.Buffer(byteBuf, 0)
+	var decodedMeta bool
+	var err error
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return err
+		}
```

Comment:
should this check be outside the `for` loop? from the docs:

```
It returns false when the scan stops, either by reaching the end of the input or an error.
```
https://pkg.go.dev/bufio#Scanner.Scan

so if `scanner.Scan()` encounters an error, it will return false and exit the loop, but the code won't have the opportunity to check the value of `scanner.Err()`?